nsls2forge_utils/auto_tick.py

```diff
@@ -0,0 +1,627 @@
+from conda_forge_tick.mamba_solver import is_recipe_solvable
+
+from .git_utils import (
+    get_repo,
+    push_repo
+)
+
+logger = logging.getLogger(__name__)
+
+PR_LIMIT = 5
+MAX_PR_LIMIT = 50
+
+MIGRATORS = [
+    Version(
+        pr_limit=PR_LIMIT * 2,
+        piggy_back_migrations=[
+            Jinja2VarsCleanup(),
+            PipMigrator(),
+            LicenseMigrator(),
+            CondaForgeYAMLCleanup(),
+            ExtraJinja2KeysCleanup(),
+        ],
+    ),
+]
+
+BOT_RERUN_LABEL = {
+    "name": "bot-rerun",
+    "color": "#191970",
+    "description": (
+        "Apply this label if you want the bot to retry "
+        "issuing a particular pull-request"
+    ),
+}
+
+
+def run(feedstock_ctx, migrator, protocol='ssh', pull_request=True,
+        rerender=True, fork=False, organization='nsls-ii-forge', **kwargs):
+    """
+    For a given feedstock and migration run the migration
+
+    Parameters
+    ----------
+    feedstock_ctx: FeedstockContext
+        The node attributes
+    migrator: Migrator instance
+        The migrator to run on the feedstock
+    protocol : str, optional
+        The git protocol to use, defaults to ``ssh``
+    pull_request : bool, optional
+        If true issue pull request, defaults to true
+    rerender : bool
+        Whether to rerender
+    fork : bool
+        If true create a fork, defaults to true
+    organization: str, optional
+        GitHub organization to get repo from
+    gh : github3.GitHub instance, optional
+        Object for communicating with GitHub, if None build from $USERNAME
+        and $PASSWORD, defaults to None
+    kwargs: dict
+        The key word arguments to pass to the migrator
+
+    Returns
+    -------
+    migrate_return: MigrationUidTypedDict
+        The migration return dict used for tracking finished migrations
+    pr_json: dict
+        The PR json object for recreating the PR as needed
+    """
+    # get the repo
+    migrator.attrs = feedstock_ctx.attrs
+
+    branch_name = migrator.remote_branch(feedstock_ctx) + "_h" + uuid4().hex[0:6]
+
+    # TODO: run this in parallel
+    feedstock_dir, repo = get_repo(
+        ctx=migrator.ctx.session,
+        fctx=feedstock_ctx,
+        branch=branch_name,
+        organization=organization,
+        feedstock=feedstock_ctx.feedstock_name,
+        protocol=protocol,
+        pull_request=pull_request,
+        fork=fork,
+
+    )
+
+    recipe_dir = os.path.join(feedstock_dir, "recipe")
+
+    # migrate the feedstock
+    migrator.run_pre_piggyback_migrations(recipe_dir, feedstock_ctx.attrs, **kwargs)
+
+    # TODO - make a commit here if the repo changed
+
+    migrate_return = migrator.migrate(recipe_dir, feedstock_ctx.attrs, **kwargs)
+
+    if not migrate_return:
+        logger.critical(
+            "Failed to migrate %s, %s",
+            feedstock_ctx.package_name,
+            feedstock_ctx.attrs.get("bad"),
+        )
+        eval_cmd(f"rm -rf {feedstock_dir}")
```

Comment:
Not sure why conda-forge team prefered to make it *nix-specific. It can be better done with https://docs.python.org/3/library/shutil.html#shutil.rmtree.